### ISSUE TITLE
IngressGateway only respects first port 443 Gateway definition

### ISSUE THREAD

**Issue by dankirkpatrick** [istio/istio#11509]
tldr: With v1.0.5, the IngressGateway only listens to the hosts provided in the first Gateway definition provided to the server.

Longer version:  I have setup the IngressGateway with a loadbalancer and a wildcard DNS entry (say, for "*.example.com").  I am trying to create 2 services/gateways/virtualservices that listen on ports 80 and 443.  For port 443, both gateways share the same wildcard TLS cert (for "*.example.com").  If I give only the explicit URL in the 443 definition for both gateways, the ingressgateway deployment only listens on the URL provided in the first gateway.  So, if the first gateway has a hosts: section with the URL "servicea.example.com", and the second gateway  has a hosts: section with the URL "serviceb.example.com", the 2nd URL will not work--the HTTPS connection will fail with the server never replying to the "hello" message sent as the first step the HTTPS connection initiation syn/ack process.  Port 80 works as expected, with the server listening to both http://servicea.example.com and http://serviceb.example.com.  But only https://servicea.example.com:443 will work--https://serviceb.example.com:443 fails to respond.

If I change the port 443 gateway definition of serviceA to include a wildcard '*' in the hosts: section, then both services work as expected.  

If I delete the gateway for serviceA, then serviceB begins to work as expected.  It then becomes the "first" gateway seen by the IngressGateway, and all other gateway URLs will fail on port 443.  Re-adding the gateway for serviceA will fail, with only the URL for serviceB listening on port 443.  Again, port 80 is unaffected by this, with the ingressgateway listening for all defijned URLs as expected.

For the record, we are using the Helm chart for v1.0.5 located in the Istio github: https://github.com/istio/istio/tree/master/install/kubernetes/helm/istio

Also, our gateway files look like this:

```apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: servicea
  namespace: default
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
      privateKey: /etc/istio/ingressgateway-certs/tls.key
    hosts:
    - servicea.example.com
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - servicea.example.com
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: serviceb
  namespace: default
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
      privateKey: /etc/istio/ingressgateway-certs/tls.key
    hosts:
    - serviceb.example.com
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - serviceb.example.com
```

The certificate and private key are for a wildcard URL "\*.example.com".  On port 80, the ingressgateway listens to both "servicea.example.com" and "serviceb.example.com".  On port 443, the ingressgateway only listens to "servicea.example.com".  Changing the hosts: value of the port 443 definition on servicea to '*' makes both services work.  Likewise, simply deleting the gateway for servicea will cause the ingressgateway to start listening on "serviceb.example.com:443".  Re-adding the gateway for servicea after this leads to only serviceb working, as it is now the "first" gateway seen by Istio.  Only the first gateway is respected.

The workaround for us to to make sure the first gateway we define has a wildcard for hosts: in the port 443 definition.  But this seems to me to be unexpected behavior.

**Comment by rsnj** [comment]
We are having the exact same issue as well with Istio 1.0.5.
It only applies to the HTTPS port as the HTTP works just fine. We had to set our first gateway to be:
```
- port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - *.example.com
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
      privateKey: /etc/istio/ingressgateway-certs/tls.key
```

Oddly, it still routes correctly to our other gateways.

**Comment by markmssd** [comment]
I'm also having the same exact issue, also with Istio 1.0.5 installed with Helm... and I also ended up having just one Gateway with wildcard, and multiple VirtualServices. Tomorrow I will try to give each Gateway a different name for port 443 (e.g. https-servicea and https-serviceb) and see if that makes a difference.

**Comment by vadimeisenbergibm** [comment]
Did you try to define a single Gateway with multiple hosts: servicea.example.com and serviceb.example.com?

**Comment by dankirkpatrick** [comment]
It is not tenable to use only one gateway.  The two services are owned by different tenants in different namespaces.  Combining the gateways will not work in our case.

**Comment by dankirkpatrick** [comment]
And, yes, if I give multiple URLs in the first gateway, it works.  Also a wildcard works in the first gateway.  Neither of these hacks is a reasonable long term solution for us, due to multi-tenancy.

**Comment by markmssd** [comment]
@vadimeisenbergibm Yes I haven't tried that no. However, here are more findings:

I defined only 1 Gateway which lists multiple port 443 (all with the same `name: https`, each for a different host. I couldn't deploy it, got an error that each port https name needs to be different.

So I tried defining multiple Gateways, each for a different host, but each having a different https name field (e.g. `name: https-servicea, name: https-serviceb`.. I think they must start with `https`, I had read it a while ago). And... it worked!!

My guess is that it's the same error was happening when we deploy multiple Gateways each with the same https names, but the error was being "swallowed" somewhere when Istio tried to merge them together.

@vadimeisenbergibm @dankirkpatrick give that a try and let me know if it works!

**Comment by dankirkpatrick** [comment]
@vadimeisenbergibm Yes, exactly.

@markmssd I just tried your solution of naming each port uniquely.  It did not work for me.

**Comment by dankirkpatrick** [comment]
@vadimeisenbergibm To comment more, different teams own different products, and deploy to different namespaces.  All of them are of the form: <product>.<company-base-url>, but each team will deploy products independently, without knowing about each other.  

As it stands now, we must define all URLs for port 443 in a single gateway, but port 80 definitions can be defined in multiple gateways.  If port 443 worked the same way as port 80, each team could deploy a Gateway/VirtualService pair for each product, and to their own namespace.  As it stands now, we either must define a wildcard in our port first port 443 gateway definition (which is what we do now), or there must be a single Gateway object that is shared by all VirtualServices.  Since each team has its own namespace, I'm not sure how that would work (I guess a fully-qualified name for the gateways field in the virtual service, along with adding an RBAC role to allow access to the gateway across namespaces?  this seems very ugly).

It seems far better to allow each team to deploy and manage their own gateways (with URLs that don't have wildcards).  There is a significant downside: two gateways deployed to different namespaces might try to use the same URL, resulting in conflicting rules.  Maybe a validating admission controller could prevent these from happening?  But this seems like a far more manageable problem than the alternatives.  Furthermore, it appears that is how port 80 already works.

**Comment by vadimeisenbergibm** [comment]
@dankirkpatrick I see, your requirements are totally reasonable.

@louiscryan @costinm @rshriram Is the gateway scoping per namespace (see 
https://github.com/istio/istio/issues/11509#issuecomment-461554163) planed to be resolved in 1.1 ?

**Comment by costinm** [comment]
I spent quite a bit of time testing this - it seems using different port names is needed with the current code, we'll need to make it very clear in the docs.

With the 'scoping' (Sidecar) API the Gateway will be defined only in the ingress namespace ( istio-system or wherever ingresses are running if using a-la-carte), so the risk of different namespaces messing it up is low.

**Comment by vadimeisenbergibm** [comment]
@costinm note that the requirement of @dankirkpatrick is to be able to define multiple Gateways in multiple namespaces by multiple teams. This is how Kubernetes Ingress works - you can define multiple Ingress resources in multiple namespaces and Kubernetes will combine them.

**Comment by dankirkpatrick** [comment]
As I've been investigating my issue, I came accross https://github.com/istio/istio/issues/5829

I think it's the same issue.  The person resolved the issue by putting a wildcard in their gateway definition.

**Comment by dankirkpatrick** [comment]
We renamed our ~20 services with unique names for each port.  That worked.  BUT...

Today two new services were rolled out with the same name for the HTTPS port.  This broke not only the 2nd gateway with the same port name, but ALL OF THE OTHER 20 SERVICES!  Having two HTTPS ports with the same name breaks other HTTPS port definitions with different names!

I would be extremely surprised if this is the expected behavior.

**Comment by costinm** [comment]
Oz, do you think Galley can validate this and reject ? 

In 1.1 we are moving to a model where Gateway resources must be located in the same namespace with the deployment - so they are easier to inspect. ( main motivation is the security of TLS certs, which have to be in the deployment namespace ).

The 'port' conflict is more of an implementation behavior we should fix.

**Comment by vadimeisenbergibm** [comment]
@costinm I think this behavior does not happen anymore after #12071.

Now you can define multiple gateways, a gateway per host. 

Let me check this and update 
https://preliminary.istio.io/docs/tasks/traffic-management/secure-ingress/mount/#configure-a-tls-ingress-gateway-for-multiple-hosts to use multiple gateways.

**Comment by vadimeisenbergibm** [comment]
@costinm @rshriram I have checked - now it is possible to define multiple gateways on the same port, a gateway per host. I rewrote the example of a secure ingress for multiple hosts - https://github.com/istio/istio.io/pull/3529.

I think we can close this issue.

**Comment by ozevren** [comment]
If I am reading this right, validation would require considering multiple resources at the same time, right?

The validation webhook can consider only one resource at a time, otherwise we risk very wonky behavior with kubectl.

**Comment by dankirkpatrick** [comment]
At the request of our Google support team, I've just completed testing this issue with Istio v1.1rc2.  The problem is still there.

My setup: I have a single ingress gateway deployment in istio-system, with a single TLS cert for a wildcard (like *.mydomain.com).  I have 17 Gateways spread across 8 namespaces.  Each namespace has unique names for the HTTP & HTTPS ports (as in, I've named the ports "https-servicename" and "http-servicename").  The desired URLS are of the form: servicename.mydomain.com.  

For HTTPS (and only HTTPS), the first gateway/virtualservice I defined using kubectl is the only URL that the ingressgateway listens to on port 443.  If I set the hosts: field for the HTTPS port definition in that first gateway to be a wildcard, all the rest of virtualservices work.  If I set the hosts: field on that first gateway to be the FQDN, the ingress gateway stops responding to any other URL on port 443 (port 80 is unaffected by this).

**Comment by vadimeisenbergibm** [comment]
@dankirkpatrick Could you please check https://preliminary.istio.io/docs/tasks/traffic-management/secure-ingress/mount/#configure-a-tls-ingress-gateway-for-multiple-hosts ? It is updated to use two gateways, a gateway per FQDN.

Do you see any differences between your setting and the one described in https://preliminary.istio.io/docs/tasks/traffic-management/secure-ingress/mount/#configure-a-tls-ingress-gateway-for-multiple-hosts ?

**Comment by costinm** [comment]
@philrud

**Comment by dankirkpatrick** [comment]
Yes, I am deploying one gateway per FQDN.

**Comment by jensendw** [comment]
I'm having the same issue with istio 1.0.6, I think the root issue here is that if you're using wildcard SSL cert you will need to share the gateway.   Initially I setup the following which didn't work:

```
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: hannibal-s1-saml-proxy
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    tls:
      httpsRedirect: true
    hosts:
    - "hannibal-s1.mydomain.com"
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
      privateKey: /etc/istio/ingressgateway-certs/tls.key
    hosts:
    - "hannibal-s1.mydomain.com"

---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: hannibal-s1-saml-proxy
spec:
  hosts:
  - "hannibal-s1.mydomain.com"
  gateways:
  - hannibal-s1-saml-proxy
  http:
  - match:
    - uri:
        prefix: /
    route:
    - destination:
        port:
          number: 80
        host: hannibal-s1-saml-proxy

---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: hannibal-p3-saml-proxy
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    tls:
      httpsRedirect: true
    hosts:
    - "hannibal-p3.mydomain.com"
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
      privateKey: /etc/istio/ingressgateway-certs/tls.key
    hosts:
    - "hannibal-p3.mydomain.com"

---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: hannibal-p3-saml-proxy
spec:
  hosts:
  - "hannibal-p3.mydomain.com"
  gateways:
  - hannibal-p3-saml-proxy
  http:
  - match:
    - uri:
        prefix: /
    route:
    - destination:
        port:
          number: 80
        host: hannibal-p3-saml-proxy
```

I switched this to one gateway definition, and individual virtual service definitions:

```
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-mydomain-internal-gateway
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    tls:
      httpsRedirect: true
    hosts:
    - "*.mydomain.com"
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
      privateKey: /etc/istio/ingressgateway-certs/tls.key
    hosts:
    - "*.mydomain.com"

---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: hannibal-p3-saml-proxy
spec:
  hosts:
  - "hannibal-p3.mydomain.com"
  gateways:
  - hannibal-p3-saml-proxy
  http:
  - match:
    - uri:
        prefix: /
    route:
    - destination:
        port:
          number: 80
        host: hannibal-p3-saml-proxy

---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: hannibal-s1-saml-proxy
spec:
  hosts:
  - "hannibal-s1.mydomain.com"
  gateways:
  - hannibal-s1-saml-proxy
  http:
  - match:
    - uri:
        prefix: /
    route:
    - destination:
        port:
          number: 80
        host: hannibal-s1-saml-proxy
```

My plan is to follow the pattern above with my other wildcard certs and have the individual application deployments reference the appropriate gateway in the virtual service they define.

I'll try to update the documentation to see if we can have a section on wildcard SSL certificates as it seems to be sort of overlooked.

**Comment by dankirkpatrick** [comment]
@jensendw That is exactly what we are doing--sharing a single wildcard cert across multiple gateways.  We will try your work-around.

This behavior we are seeing sounds slightly different than #12500.  This affects gateways with distinct port names, as well.  

Is there any reason multiple gateways cannot share the same wildcard cert?  By having a single Gateway definition across all of our VirtualServices, we require different tenants in different namespaces to write to a shared object, the Gateway....or, alternatively, having a wildcard in the Gateway.  I don't think either of these are acceptable long-term work-arounds.

**Comment by philrud** [comment]
@dankirkpatrick the root cause is the same. I got a PR that fixes #12500 and I've tested a wildcard cert shared across two namespaces/gateways, combination of wildcard+normal cert, etc. -- that works. I believe it's too late for it to go to 1.1, but 1.1.1 should be possible

**Comment by duderino** [comment]
Fixed by #12556 which will be in 1.1.1
